bitcoin/src/address.rs

```diff
@@ -565,22 +632,75 @@ impl<'a> fmt::Display for AddressEncoding<'a> {
 /// * [BIP341 - Taproot: SegWit version 1 spending rules](https://github.com/bitcoin/bips/blob/master/bip-0341.mediawiki)
 /// * [BIP350 - Bech32m format for v1+ witness addresses](https://github.com/bitcoin/bips/blob/master/bip-0350.mediawiki)
 #[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
-pub struct Address {
+pub struct Address<V = NetworkChecked>
+where
+    V: NetworkValidation,
+{
     /// The type of the address.
     pub payload: Payload,
     /// The network on which this address is usable.
     pub network: Network,
+    /// Marker of the status of network validation.
+    _validation: PhantomData<V>,
 }
+
+#[cfg(feature = "serde")]
+crate::serde_utils::serde_string_impl!(Address<NetworkUnchecked>, "a Bitcoin address");
+
+// Macro serde_string_impl creates Serialize and Deserialize for Address<NetworkUnchecked>
+// Here we add Serialize for Address<NetworkChecked>. There is no Deserialize for it.
 #[cfg(feature = "serde")]
-crate::serde_utils::serde_string_impl!(Address, "a Bitcoin address");
+impl serde::Serialize for Address {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        serializer.collect_str(&self)
+    }
+}
 
+/// Methods on [`Address`] that can be called on both `Address<NetworkChecked>` and
+/// `Address<NetworkUnchecked>`.
+impl<V: NetworkValidation> Address<V> {
+    /// Gets the address type of the address.
+    ///
+    /// # Returns
+    /// None if unknown, non-standard or related to the future witness version.
+    pub fn address_type(&self) -> Option<AddressType> {
+        match self.payload {
+            Payload::PubkeyHash(_) => Some(AddressType::P2pkh),
+            Payload::ScriptHash(_) => Some(AddressType::P2sh),
+            Payload::WitnessProgram { version, program: ref prog } => {
+                // BIP-141 p2wpkh or p2wsh addresses.
+                match version {
+                    WitnessVersion::V0 => match prog.len() {
+                        20 => Some(AddressType::P2wpkh),
+                        32 => Some(AddressType::P2wsh),
+                        _ => None,
+                    },
+                    WitnessVersion::V1 if prog.len() == 32 => Some(AddressType::P2tr),
+                    _ => None,
+                }
+            }
+        }
+    }
+
+    /// Create new address from given components, infering the network validation
+    /// marker type of the address.
+    #[inline]
+    fn new(network: Network, payload: Payload) -> Address<V> {
+	Address { network, payload, _validation: PhantomData }
+    }
+}
+
+/// Methods and functions on [`Address`] that can be called only on `Address<NetworkValid>`.
```

Comment:
This line is stale, should be
```suggestion
/// Methods and functions that can be called only on `Address<NetworkChecked>`.
```

Reply:
Right, thanks!